### PR TITLE
adds support for timeout and custom image in cloudshell

### PR DESCRIPTION
**Pull request by invincibleJai**
**Fixes**: 
https://issues.redhat.com/browse/ODC-7197

**Solution Description**: 

adds support for timeout and custom image in the cloud shell

**Screenshots / Gifs for design review**: 

- Admin

wto-form-admin-2

image

image

- Developer
wto-dev-3

image

Test setup:
- (for local dev setup follow https://docs.google.com/document/d/173PFbaTMWHf8PAhXmWomAaSbxmVwO2Dpal_qBpqxLe8/edit#)
- Install WTO operator
- Click on the terminal icon top right header
- Test as admin/self-provisioner as well

**TODO**: 
- [x] add unit tests
- [x] add e2e tests for kafka sink creation

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Review by amisevsk — COMMENTED** [review]
Looks great to me so far. A few notes:

* The hint for timeout describes the string format (`5m, 1.5h`) but the input format doesn't match (number + time unit selector). Maybe we should update this or remove the format hint.
* Defaults are somewhat unclear -- I realize it's likely not possible to read the actual default values (especially image which changes relatively often), but once the panel for timeout is opened, it's unclear how to _not_ set a custom value. I assume that if the image field is left blank, the image is unset and the default is used, but for timeout the field is initially populated with 0 -- how do I leave it unset once the advanced option is opened? What happens if `0` is left as the value when the workspace is started?
    * This can perhaps be worked-around by pre-filling the timeout field with the current default value (`15m`) but this would be fragile since this default can be configured.
* In the demo recordings, it appears that the custom image is potentially not being used. `quay.io/amisevsk/web-terminal-tooling:testing` is a custom tooling image that should print the following on loading the terminal:
    ```
    Welcome to the OpenShift Web Terminal. Type "help" for a list of installed CLI tools.
    This is a custom image pulled from quay.io/amisevsk/web-terminal-tooling:testing
    ```

As a side note, you should be able to use the `wtoctl` CLI utility within the running terminal to check that values were set: `wtoctl get timeout` will print the currently configured idle timeout,and `wtoctl get image` will get the currently configured image.

**Review by amisevsk — COMMENTED** [review]
Tested via cluster-bot, works as expected generally :+1:

Apart from other comments above, I noticed that if you start a terminal with timeout set to `0`, the DevWorkspace gets an extra empty section:

```yaml
spec:
  template:
    components:
    # ...
    - name: web-terminal-exec
      plugin:
        components:               # unnecessary
        - container: {}           # unnecessary
          name: web-terminal-exec # unnecessary
        kubernetes:
          name: web-terminal-exec
          namespace: openshift-operators
```

File: frontend/packages/console-shared/locales/en/devconsole.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "Click on the names to access advanced options for <0></0>.": "Click on the names to access advanced options for <0></0>."
+}
```
Comment (vikram-raj) [inline frontend/packages/console-shared/locales/en/devconsole.json:3]: Looks like a wrong translation file.

File: frontend/packages/console-shared/src/components/progressive-list/ProgressiveListItem.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.ocs-progressive-list-section {
+  margin-bottom: var(--pf-global--spacer--md);
+}
```
Comment (vikram-raj) [inline frontend/packages/console-shared/src/components/progressive-list/ProgressiveListItem.scss:3]: nit: add new line.

File: frontend/packages/dev-console/src/components/progressive-list-bkp/ProgressiveList.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        <React.Fragment key={item}>
+          {validChildren.find(({ props }: React.ReactElement) => item === props.name)}
+        </React.Fragment>
+      ))}
+      <ProgressiveListFooter Footer={Footer} items={items} onShowItem={onVisibleItemChange} />
+    </>
+  );
+};
+
+export default ProgressiveList;
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/progressive-list-bkp/ProgressiveList.tsx:39]: nit: looks like you missed to delete `progressive-list` component from the `dev-console` package.

**Review by vikram-raj — COMMENTED** [review]
The overall feature works fine. Added a few comments.

- footer width is not 100%
image

- When I change the unit without touching the number field, I got the below error, and the error message is not user-friendly. 
image

File: frontend/packages/console-app/src/components/cloud-shell/setup/CloudShellAdminSetup.tsx [hunk 4]
```diff
@@ -1,71 +1,100 @@
...
+      timeout: {
+        limit: null,
+        unit: 'm',
+      },
+    },
+  };
+
   const { t } = useTranslation();
 
   const [initError, setInitError] = React.useState<string>();
```
Comment (jerolimov) [inline frontend/packages/console-app/src/components/cloud-shell/setup/CloudShellAdminSetup.tsx:47]: Also if this is old code, I don't see any place where a string was saved here.

```suggestion
  const [initError, setInitError] = React.useState<Error>();
```

**Review by vikram-raj — COMMENTED** [review]
All looks good to me. The only issue I noticed is sticky footer width is not 100%.

image

File: frontend/packages/console-app/src/components/cloud-shell/setup/CloudShellAdminSetup.tsx (hunk 4 above)
Comment (invincibleJai) [reply frontend/packages/console-app/src/components/cloud-shell/setup/CloudShellAdminSetup.tsx:47]: yeah it's an existing one and `initError` is passed here https://github.com/openshift/console/blob/master/frontend/packages/console-app/src/components/cloud-shell/setup/CloudShellAdminSetup.tsx#L72

`<LoadError message={initError} label={t('console-app~OpenShift command line terminal')} />`

so ideally should keep as a string only and pass `error.message`

WDYT?

**Comment by invincibleJai** [comment]
> All looks good to me. The only issue I noticed is sticky footer width is not 100%.
> 
> image

Thanks Vikram, have fixed this.

**Review by amisevsk — APPROVED** [review]
Tested changes and looks good to me :+1:

**Review by vikram-raj — APPROVED** [review]
Verified it and it works as expected. Thanks @invincibleJai 

/lgtm

**Comment by invincibleJai** [comment]
/label docs-approved
/label px-approved

Propagating from epic

**Comment by invincibleJai** [comment]
/cc @sanketpathak 

PTAL when you get time, thanks!!

**Comment by sanketpathak** [comment]
Verified it on a cluster using cluster-bot 
Screenshot from 2022-12-19 22-35-33
/label qe-approved

File: frontend/packages/console-app/locales/en/console-app.json [hunk 5]
```diff
@@ -223,13 +223,22 @@
...
   "Cancel": "Cancel",
+  "Set custom image for the terminal.": "Set custom image for the terminal.",
   "Select Project": "Select Project",
   "Create Project": "Create Project",
-  "This Project will be used to initialize your command line terminal": "This Project will be used to initialize your command line terminal",
   "Project name": "Project name",
+  "Seconds": "Seconds",
+  "Minutes": "Minutes",
+  "Hours": "Hours",
+  "Miliseconds": "Miliseconds",
```
Comment (amisevsk) [inline frontend/packages/console-app/locales/en/console-app.json:240]: `s/Miliseconds/Milliseconds/`

**Review by amisevsk — COMMENTED** [review]
Just noticed a typo in this PR on an old tab, sorry I missed it -- cc: @invincibleJai

File: frontend/packages/console-app/locales/en/console-app.json (hunk 5 above)
Comment (invincibleJai) [reply frontend/packages/console-app/locales/en/console-app.json:240]: Thaks @amisevsk, here is the PR to get this right https://github.com/openshift/console/pull/12423

**Pull request merged**